xilem_core/src/views/run_once.rs

```diff
@@ -0,0 +1,119 @@
+// Copyright 2024 the Xilem Authors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{DynMessage, MessageResult, NoElement, View, ViewPathTracker};
+
+/// A view which executes `once` exactly once.
+///
+/// `once` will be called only when the returned view is [built](View::build).
+///
+/// This is a [`NoElement`] view, and so should either be used in any sequence, or with [`fork`](crate::fork).
+///
+/// ## Examples
+///
+/// This can be useful for logging a value:
+///
+/// ```
+/// # use xilem_core::{run_once, View, docs::{Fake as ViewCtx}, PhantomView};
+/// # struct AppData;
+/// fn log_lifecycle(data: &mut AppData) -> impl PhantomView<AppData, (), ViewCtx> {
+///     run_once(|| eprintln!("View constructed"))
+/// }
+/// ```
+/// ## Capturing
+///
+/// This method cannot be used with a dynamic `once`.
+/// That is, `once` cannot be a function pointer or capture any (non-zero sized) values.
+/// You might otherwise expect the function to be reran when the captured values change, which is not the case.
+/// [`run_once_raw`] is the same as `run_once`, but without this restriction.
+///
+/// // https://doc.rust-lang.org/error_codes/E0080.html
+/// // Note that this error code is only checked on nightly
+/// ```compile_fail,E0080
+/// # use xilem_core::{run_once, View, docs::{Fake as ViewCtx}, PhantomView};
+/// # struct AppData {
+/// #    data: u32
+/// # }
+/// fn log_data(app: &mut AppData) -> impl PhantomView<AppData, (), ViewCtx> {
+///     let val = app.data;
+///     run_once(move || println!("{}", val))
+/// }
+/// # // We need to call the function to make the inline constant be evaluated
+/// # let _ = log_data(&mut AppData { data: 10 });
+/// ```
+pub fn run_once<F>(once: F) -> RunOnce<F>
+where
+    F: Fn() + 'static,
```

Comment:
I could also imagine this taking the mutable app state, triggered via a message that is sent to this view, but I guess this could also be a separate view, and maybe even take an `async` function then. Something like [this](https://github.com/Philipp-M/trui/blob/dd24a2921eb0587dfffbbafea4a3e62b92f39c43/examples/async_event_handler_stream.rs#L67) (in that context the `EventHandler` trait introduced there also gets more interesting again...)

Reply:
As in make this a view which sends a message to itself?
It's an interesting idea, although it makes things more complicated; it means that the `ctx` needs to be able to send itself a message, which feels a bit weird.

This taking an async function is kind of missing the point of this view, which is that it can exist in `xilem_core`.

Reply:
Yeah as said, this is a different kind of view, this view is fine to me, I guess it's more for debugging purposes?

>  it means that the ctx needs to be able to send itself a message, which feels a bit weird.

Is it though? We don't have access to the app state in `build` or `rebuild` so we kind of need to do it in `View::message`, when we want access to app state.

> This taking an async function is kind of missing the point of this view, which is that it can exist in xilem_core.

Good point

Reply:
I don't think that new capability belongs in this PR. You can use `run_once_raw` to *read* the app state, and it feels like it's asking for infinite loops to modify the app state in response to creating a view.

I'm hopeful that we can avoid needing this entirely.

Reply:
Yeah, I think we can postpone all of this until it really will be necessary or brings a "measurable" benefit.

Though I don't think a `RunOnce` view with access to the app state should function like this view (i.e. it should rather send a message to itself, in build, so that it can "cleanly" access app state in `View::message` and change it, but as you say, an infinite loop could be created anyway. I think `run_once_raw` can be removed then?